lib/backend/etcdv3/etcdv3.go

```diff
@@ -0,0 +1,462 @@
+
+	return d, nil
+}
+
+// Update an entry in the datastore.  If the entry does not exist, this will return
+// an ErrorResourceDoesNotExist error.  The ResourceVersion must be specified, and if
+// incorrect will return a ErrorResourceUpdateConflict error and the current entry.
+func (c *EtcdV3Client) Update(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"model-key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Update request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+	logCxt = logCxt.WithField("etcdv3-key", key)
+
+	opts, err := c.getTTLOption(d)
+	if err != nil {
+		return nil, err
+	}
+
+	// ResourceVersion must be set for an Update.
+	rev, err := strconv.ParseInt(d.Revision, 10, 64)
+	if err != nil {
+		logCxt.Info("Unable to parse Revision")
+		return nil, err
+	}
+	conds := []clientv3.Cmp{clientv3.Compare(clientv3.ModRevision(key), "=", rev)}
+
+	logCxt.Debug("Performing etcdv3 transaction for Update request")
+	txnResp, err := c.etcdClient.Txn(context.Background()).If(
+		conds...,
+	).Then(
+		clientv3.OpPut(key, value, opts...),
+	).Else(
+		clientv3.OpGet(key),
+	).Commit()
+
+	if err != nil {
+		logCxt.WithError(err).Warning("Update failed")
+		return nil, errors.ErrorDatastoreError{Err: err}
+	}
+
+	// Etcd V3 does not return a error when compare condition fails we must verify the
+	// response Succeeded field instead.  If the compare did not succeed then check for
+	// a successful get to return either an UpdateConflict or a ResourceDoesNotExist error.
+	if !txnResp.Succeeded {
+		getResp := (*clientv3.GetResponse)(txnResp.Responses[0].GetResponseRange())
+		if len(getResp.Kvs) == 0 {
+			logCxt.Info("Update transaction failed due to resource not existing")
+			return nil, errors.ErrorResourceDoesNotExist{Identifier: d.Key}
+		}
+
+		logCxt.Info("Update transaction failed due to resource update conflict")
+		var existing *model.KVPair
+		if v, err := model.ParseValue(d.Key, getResp.Kvs[0].Value); err == nil {
+			logCxt.Debug("Parsed current entry to return in response")
+			existing = &model.KVPair{
+				Key:      d.Key,
+				Value:    v,
+				Revision: strconv.FormatInt(getResp.Kvs[0].ModRevision, 10),
+			}
+		}
+		return existing, errors.ErrorResourceUpdateConflict{Identifier: d.Key}
+	}
+
+	d.Revision = strconv.FormatInt(txnResp.Header.Revision, 10)
+
+	return d, nil
+}
+
+// Apply an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdV3Client) Apply(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Apply request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+
+	logCxt.Debug("Performing etcdv3 Put for Apply request")
+	resp, err := c.etcdClient.Put(context.Background(), key, value)
```

Comment:
for `Apply` we usually call `Update` & `Create`, why haven't we done the same here?

Reply:
another reason being `apply` = `create` + `update` so we want to use the same options we used in create and update, and if we change one of those we want `apply` to change with it

Reply:
I am planning to remove the Apply  and just support Create/Update.

For calicoctl, Apply will need additional logic to ensure we don't lose the read-only fields in the Metadata.